cpp/include/cudf/strings/regex/config.hpp

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/regex/flags.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+
+namespace cudf {
+namespace strings {
+
+/**
+ * @addtogroup strings_regex
+ * @{
+ */
+
+/**
+ * @brief Compute the working memory size for evaluating a regex pattern
+ * on a given strings column.
+ *
+ * This function returns the size in bytes of the memory needed to evaluate
+ * the given regex pattern in parallel over the returned output rows.
+ * The number of output rows will be less than or equal to the size of the
+ * input column.
+ *
+ * This function computes only the state data memory size required to process
+ * a regex pattern over the output row count.
+ * Specific functions that use regex may require additional working memory
+ * unrelated to the regex processing.
+ *
+ * @param input Strings instance
+ * @param pattern Regex pattern to be used
+ * @param flags Regex flags for interpreting special characters in the pattern
+ * @return Size of the state memory in bytes required for processing `pattern` on `strings`
+ *         and the number of concurrent rows this memory will support
+ */
+std::pair<std::size_t, size_type> compute_regex_state_memory(
```

Comment:
This API seems fine for a "just in time" check for how much memory the call will take, but currently the RAPIDS Accelerator does "up front" planning on the driver node (which does not necessarily have a GPU).  At that point we don't know what the column data will be, just the pattern that will be used.

Is there a way to get a worst-case ballpark estimate of what memory will be used based solely on the pattern?  Alternatively, it would be great if we could specify a maximum amount of memory we want the regex to use internally, and the concurrent rows would shrink to fit within that memory (or fail if it cannot).

Reply:
Right now (with this API) you could call it with 1 row and do the math to help estimate the max size. 
I'm trying to avoid another API to set the size unless absolutely necessary.

Reply:
But doesn't the row require having a device?  We cannot require the driver node has a device.

Reply:
> But doesn't the row require having a device? We cannot require the driver node has a device.

This API does requires a device.

Reply:
Currently the API requires a string_view, but only to get the number of rows within it.  Can we just pass the number of rows directly along with the pattern rather than a type that requires a device?

Reply:
I'd rather not. In the future the state memory size may actually be influenced by the string data (e.g. the max string width) so this would likely require a kernel call.

Reply:
OK.  This will be an API the RAPIDS Accelerator will not be able to use until it can do a late, dynamic fallback to the Spark CPU implementation for a regex expression (converting columnar to row and back to columnar in the process).

Reply:
Sorry I did not  understand the requirements for this API. Should I close this PR?

Reply:
> Should I close this PR? 

Possibly, if the RAPIDS Accelerator is the only use-case for it.

It does look useful for the "just in time" check to avoid using excessive memory.  Unfortunately we currently cannot fallback to the CPU for regex expressions that late in the processing.  The RAPIDS Accelerator performs up-front planning of the query on the Spark driver, which does not have access to a GPU, to decide which parts of the query will or will not be on the GPU during execution.  Once the driver decides, that's how we're going to run (i.e.: no late data-dependent fallback yet).

What we'd like is either the ability to estimate memory usage based solely on the pattern string (without requiring a GPU to do the estimate) or the ability to control how much memory will be used during regex processing (potentially reducing concurrency to fit).

Reply:
The actual memory size is going to be based on the string data as well as the pattern. At best, we could give a maximum as a worst case based only on the pattern. If you could create an issue with your requirements (including the _no GPU device_ detail) then I think we can close this PR and prioritize that work.

Reply:
I filed #10852 for the API we would ultimately like, which is the ability to control the temporary memory size.  Given we cannot leverage an accurate memory reporting API because of the need to decide at query planning time and lack of knowledge of the specific search data in the batch, any memory size reporting API is going to be a guesstimate that could be way off in the actual case.

We already have to walk the regex operations in the pattern to transpile it to a cudf compatible regex, and we could put some heuristics there to avoid GPU processing of crazy large patterns etc. as a stop-gap measure until we have the ability to budget and control the regex memory usage.

It occurred to me that we can still leverage this API as it is.  We cannot use it on the driver to make the decision to fallback, but we can use it on the executor to decide if we need to slice the input into smaller batches to make everything fit.  If the reported memory is going to be "too big" we could slice the input column into smaller batches until each batch will fit, process the batches serially, then concatenate the results back into the final result.  We will have to do some estimating on where to slice the input, but this might be effective in practice.  Does that seem like a workable approach?

Reply:
Yes, that sounds reasonable.

Reply:
I really don't see us using this in the current form. We can keep the code around on this branch in case things change and we have to resurrect it. But I don't see a reason to check it in or keep the PR open.